src/configureStore.ts

```diff
@@ -0,0 +1,140 @@
+import {
+  createStore,
+  compose,
+  applyMiddleware,
+  combineReducers,
+  Reducer,
+  ReducersMapObject,
+  Middleware,
+  Action,
+  AnyAction,
+  StoreEnhancer,
+  Store,
+  DeepPartial
+} from 'redux'
+import { composeWithDevTools } from 'redux-devtools-extension'
+import thunk from 'redux-thunk'
+import createImmutableStateInvariantMiddleware from 'redux-immutable-state-invariant'
+import createSerializableStateInvariantMiddleware from './serializableStateInvariantMiddleware'
+
+import isPlainObject from './isPlainObject'
+
+const IS_PRODUCTION = process.env.NODE_ENV === 'production'
+
+/**
+ * Returns any array containing the default middleware installed by
+ * `configureStore()`. Useful if you want to configure your store with a custom
+ * `middleware` array but still keep the default set.
+ *
+ * @return The default middleware used by `configureStore()`.
+ */
+export function getDefaultMiddleware(
+  isProduction = IS_PRODUCTION
+): Middleware[] {
```

Comment:
Doing it like this

```js
const IS_PRODUCTION = process.env.NODE_ENV === 'production';

export function getDefaultMiddleware(isProduction = IS_PRODUCTION) {
  const middlewareArray = [thunk,];
  let middlewareArrayPlus;

  if (!isProduction) {
    middlewareArrayPlus = [
      createImmutableStateInvariantMiddleware(),
      thunk,
      createSerializableStateInvariantMiddleware(),
    ];
  }

  return middlewareArrayPlus === undefined
    ? middlewareArray
    : middlewareArrayPlus;
}
```
preserves the types

Reply:
Hm, yeah, this could be `[ThunkMiddleware<S, A>, ...Middleware[]]` for instance. However I have no idea how we would preserve the extension type in `configureStore`. Would we need overrides with a one-element `middleware` tuple, two-element `middleware` tuple, etc., so that we can write `Store<S,A> & Ext1 & Ext2 & ...`?

Reply:
yeah, we would need to write an exponential amount of overloads since not only is the number of `middlewares` variable the number of `enhancers` is as well.
So we'd end up with overrides for one-element `middleware` tuple and one-element `enhancer` tuple, one-element `middleware` tuple and two-element `enhancer` tuple, two-element middleware tuple and one-element `enhancer` tuple, two-element `middleware` tuple and two-element `enhancer` tuple, etc

Reply:
Yes, I think this is a point where we need to favor practicality over maximum type precision. If we cannot find an economical way to get the type just right, we'll probably get by just fine by having the configured store's `dispatch` accept a relatively generic type like `unknown` for now. We can later see if we can tweak the API to make typing easier.

Reply:
For now, I addressed the problem by letting `configureStore()` return an `EnhancedStore extends Store` type that is hardcoded to allow dispatching thunks in addition to plain actions. This at least unblocks the use of redux-thunk. Let's improve on this after this PR (perhaps with the help of an API change).